packages/core/src/match.ts

```diff
@@ -0,0 +1,143 @@
+
+  if (filter.value === '' && values.length === 0) {
+    // If the filter operator is "equals", then the filter matches.
+    // If the filter operator is "not equals", then the filter does not match.
+    return filter.operator === Operator.EQUALS;
+  }
+
+  // Normalize the values array into reference strings
+  const references = values.map((value) => (typeof value === 'string' ? value : value.reference));
+
+  for (const filterValue of filter.value.split(',')) {
+    const found = references.includes(filterValue);
+    if (filter.operator === Operator.NOT_EQUALS ? !found : found) {
+      return true;
+    }
+  }
+  return false;
+}
+
+function matchesTokenFilter(resource: Resource, filter: Filter, searchParam: SearchParameter): boolean {
+  const details = getSearchParameterDetails(resource.resourceType, searchParam);
+  if (details.type === SearchParameterType.BOOLEAN) {
+    return matchesBooleanFilter(resource, filter, searchParam);
+  } else {
+    return matchesStringFilter(resource, filter, searchParam);
+  }
+}
+
+function matchesBooleanFilter(resource: Resource, filter: Filter, searchParam: SearchParameter): boolean {
+  const values = evalFhirPath(searchParam.expression as string, resource);
+  const expected = filter.value === 'true';
+  const result = values.includes(expected);
+  return filter.operator === Operator.NOT_EQUALS ? !result : result;
+}
+
+function matchesStringFilter(resource: Resource, filter: Filter, searchParam: SearchParameter): boolean {
+  const resourceValues = evalFhirPath(searchParam.expression as string, resource);
+  const filterValues = filter.value.split(',');
+  for (const resourceValue of resourceValues) {
+    for (const filterValue of filterValues) {
+      if (matchesStringValue(resourceValue, filter.operator, filterValue)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+function matchesStringValue(resourceValue: unknown, operator: Operator, filterValue: string): boolean {
+  let str = '';
+  if (resourceValue) {
+    if (typeof resourceValue === 'string') {
+      str = resourceValue;
+    } else if (typeof resourceValue === 'object') {
+      str = JSON.stringify(resourceValue);
+    }
+  }
+
+  const isMatch = str.toLowerCase().includes(filterValue.toLowerCase());
+  return operator === Operator.NOT_EQUALS ? !isMatch : isMatch;
+}
+
+function matchesDateFilter(resource: Resource, filter: Filter, searchParam: SearchParameter): boolean {
+  const resourceValues = evalFhirPath(searchParam.expression as string, resource);
+  const filterValues = filter.value.split(',');
+  for (const resourceValue of resourceValues) {
+    for (const filterValue of filterValues) {
+      if (matchesDateValue(resourceValue as string, filter.operator, filterValue)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+function matchesDateValue(resourceValue: string, operator: Operator, filterValue: string): boolean {
+  switch (operator) {
+    case Operator.GREATER_THAN:
```

Comment:
Do we need to handle the filters like "Ends Before", "Starts After" here?

Reply:
✅